cis194/week7/daniel-deng/src/CIS194/Homework07/Exercise04.hs

```diff
@@ -0,0 +1,49 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module CIS194.Homework07.Exercise04 where
+
+import CIS194.Homework07.Buffer     ( Buffer(..) )
+import CIS194.Homework07.Scrabble   ( Score(..), getScore )
+import CIS194.Homework07.Sized      ( Size(..), getSize, size )
+import CIS194.Homework07.JoinList   ( JoinList(..), tag )
+import CIS194.Homework07.Exercise02 ( indexJ, dropJ, takeJ )
+import CIS194.Homework07.Exercise03 ( scoreString )
+
+type JoinListBuffer = JoinList (Score, Size) String
+
+jlToList :: Monoid m => JoinList m a -> [a]
+jlToList Empty              = []
+jlToList (Single _ a)       = [a]
+jlToList (Append _ jl1 jl2) = jlToList jl1 ++ jlToList jl2
+
+instance Buffer JoinListBuffer where
+
+  -- | Convert a buffer to a String.
+  toString = unlines . jlToList
+
+  -- | Create a buffer from a String.
+  fromString = mconcat . fmap toSingle . lines
+             where
+               toSingle = flip Single <*> annotations
+               annotations = (,) . scoreString <*> const (Size 1)
```

Comment:
@shineli1984 would you personally do applicative style composition here? Do you find it harder to read than the _normal code_ ?

```suggestion
               annotations s = (scoreString s,  Size 1)
```

Reply:
I wouldn’t  because we are adding const just for the sake of using <*>

Reply:
@shineli1984 thanks! https://github.com/fp-works/2019-winter-Haskell-school/pull/117/commits/c57fd965679c6f053914c48709128a04b6c0209f